### PR TITLE
 #1157 notify all interested clients when a chain eviction occurs on the server

### PR DESCRIPTION
**Pull request by lorban**

### CODE REVIEW

File: clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java [hunk 1]
```diff
@@ -28,17 +31,48 @@
...
 
 class OffHeapChainMap<K> implements MapInternals {
 
+  interface ChainMapEvictionListener<K> {
+    void onEviction(K key);
+  }
+
   private final ReadWriteLockedOffHeapClockCache<K, InternalChain> heads;
   private final OffHeapChainStorageEngine chainStorage;
+  private final List<ChainMapEvictionListener<K>> evictionListeners = new CopyOnWriteArrayList<ChainMapEvictionListener<K>>();
```
Comment (ljacomet) [inline clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java]: Why not a single one? Do we really need multiple here?

File: clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java [hunk 2]
```diff
@@ -29,14 +32,29 @@
 public class OffHeapServerStore implements ServerStore {
 
   private final List<OffHeapChainMap<Long>> segments;
+  private final List<ServerStoreEvictionListener> listeners = new CopyOnWriteArrayList<ServerStoreEvictionListener>();
```
Comment (ljacomet) [inline clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java]: Why not a single one? Seeing other places, it looks like we stack three collections on top of each other - that seems overkill.

File: clustered/common/src/main/java/org/ehcache/clustered/common/messages/ResponseCodec.java [hunk 3]
```diff
@@ -94,6 +96,15 @@
         buffer.putInt(((ClientInvalidateAll) response).getInvalidationId());
         return buffer.array();
       }
+      case SERVER_INVALIDATE_HASH: {
+        ServerInvalidateHash serverInvalidateHash = (ServerInvalidateHash) response;
+        byte[] cacheIdBytes = serverInvalidateHash.getCacheId().getBytes(Charset.forName("UTF-8"));
```
Comment (ljacomet) [inline clustered/common/src/main/java/org/ehcache/clustered/common/messages/ResponseCodec.java]: We have inconsistent cache id encoding between this class and `ServerStoreOpCodec`. Would be good to pick one and only one.

File: clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java (hunk 2 above)
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java]: You're right, there will never be a need for multiple listeners. Let me rationalize all that.

File: clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java (hunk 1 above)
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java]: Nope, you're right.

File: clustered/common/src/main/java/org/ehcache/clustered/common/messages/ResponseCodec.java (hunk 3 above)
Comment (lorban) [inline clustered/common/src/main/java/org/ehcache/clustered/common/messages/ResponseCodec.java]: I'm going to commonize the code that encodes strings, but the more I look at the message / codec code the more I think this should all be handled with something like Protobuf.

**Pull request merged**